src/core/manager.c

```diff
@@ -1019,6 +1021,53 @@ int manager_new(RuntimeScope runtime_scope, ManagerTestRunFlags test_run_flags,
 
                 if (r < 0 && r != -EEXIST)
                         return r;
+
+                m->executor_fd = open(SYSTEMD_EXECUTOR_BINARY_PATH, O_CLOEXEC|O_PATH);
+                if (m->executor_fd < 0)
+                        return log_warning_errno(errno,
+                                                 "Failed to open executor binary '%s': %m",
+                                                 SYSTEMD_EXECUTOR_BINARY_PATH);
+
+                /* posix_spawn cannot take a FD, so pin the path */
+                m->executor_path = strdup(FORMAT_PROC_FD_PATH(m->executor_fd));
+                if (!m->executor_path)
+                        return -ENOMEM;
+        } else if (!FLAGS_SET(test_run_flags, MANAGER_TEST_DONT_OPEN_EXECUTOR)) {
+                _cleanup_free_ char *self_exe = NULL, *self_dir = NULL, *executor_path = NULL;
+                _cleanup_close_ int self_dir_fd = -EBADF;
+
+                /* Prefer sd-executor from the same directory as the test, e.g.: when running unit tests from the
+                * build directory. Fallback to working directory and then the installation path. */
+                r = readlink_and_make_absolute("/proc/self/exe", &self_exe);
+                if (r < 0)
+                        return r;
+
+                r = path_extract_directory(self_exe, &self_dir);
+                if (r < 0)
+                        return r;
+
+                self_dir_fd = open(self_dir, O_CLOEXEC|O_DIRECTORY);
+                if (self_dir_fd < 0)
+                        return -errno;
+
+                m->executor_fd = openat(self_dir_fd, "systemd-executor", O_CLOEXEC|O_PATH);
+                if (m->executor_fd < 0 && errno == ENOENT)
+                        m->executor_fd = openat(AT_FDCWD, "systemd-executor", O_CLOEXEC|O_PATH);
+                if (m->executor_fd < 0 && errno == ENOENT)
+                        m->executor_fd = open(SYSTEMD_EXECUTOR_BINARY_PATH, O_CLOEXEC|O_PATH);
+                if (m->executor_fd < 0)
+                        return -errno;
+
```

Comment:
add a debug log message here about the binary used (and maybe even raise to LOG_WARNING level if this is the system executor binary, not the local one)